tests/lib/rules/no-loop-func.js

```diff
@@ -111,7 +111,65 @@ ruleTester.run("no-loop-func", rule, {
                 "let a;"
             ].join("\n"),
             parserOptions: { ecmaVersion: 6 }
+        },
+
+        /*
+         * These loops _look_ like they might be unsafe, but because i is undeclared, they're fine
+         * at least as far as this rule is concerned - the loop doesn't declare/generate the variable.
+         */
+        {
+            code: "while(i) { (function() { i; }) }",
+            errors: [{ messageId: "unsafeRefs", data: { varNames: "'i'" }, type: "FunctionExpression" }]
+        },
+        {
+            code: "do { (function() { i; }) } while (i)",
+            errors: [{ messageId: "unsafeRefs", data: { varNames: "'i'" }, type: "FunctionExpression" }]
+        },
+
+        /**
+         * These loops _look_ like they might be unsafe, but because i is declared outside the loop,
+         * they're fine as far as this rule is concerned - the loop doesn't declare/generate the variable.
+         * The variable that's captured is just the one variable shared by all the loops, but that's
+         * explicitly expected in these cases.
+         */
```

Comment:
This description might be misleading. The references to `i` are considered safe because `i` isn't updated in or after the loop. For example, these would be unsafe:

```js
var i; while(i) { (function() { i; }); i++ }

var i; while(i) { (function() { i; }) } i++
```

Reply:
I'd disagree that those are examples of unsafe usage, at least as defined by the rule.

My understanding is that the rule is looking for variables that _look_ like they are scoped solely to the loop and are a new variable on each iteration, but aren't because of how `var` gets scoped.

i.e. it's trying to point out to devs that `for (var i = 0; i < 10; i++) { () => i; }` is actually equivalent to `var i = 0; for (; i < 10; i++) { () => i; }` because of how `var` scopes.

In the two examples you give, there's no ambiguity on the scope of `i` - it's declared outside the loop, so there's no potential confusion as to whether it is scoped to the loop or outside it - it's clearly and explicitly a single variable shared by all iterations of the loop.

As a contrived counter-example that uses/captures/mutates a variable in all the weird positions, but is still safe:

````js
function getLoggers(defaultLogLevel) {
    const sevs = ['error', 'info', 'debug', 'none'];
    var logLevel = 'error';
    const logFuncs = {
        setLevel(level) { logLevel = level; }
    };
    while (logLevel !== 'none') { // as a condition of the loop
        Object.defineProperty(logFuncs, logLevel, { value: (function(message) {
            if (this.logLevel === logLevel) { // captured in the loop
                console.log(message);
            }
        }).bind( { logLevel }); // used in the loop
        logLevel = sevs[sevs.indexOf(logLevel) + 1]; // mutated in the loop
    }

    // Set default log level
    logLevel = defaultLogLevel; // mutated after the loop
    return logFuncs;
}
```

`logFuncs.info('blah')` will behave as expected, only writing to the console after a `setLevel('info')` call, so the capture is safe, despite the various mutations. But I would not expect `no-loop-func` to warn about the capture in the loop in this case.

Reply:
Looking at the test cases quickly, it does appear that the rule would currently error for your examples cases and my contrived counter-example. I'd propose that those cases be reviewed - I don't think it should be warning on those, as the usages are safe, and the scoping isn't unclear.

Reply:
@matwilko let's open a separate issue for that to keep this PR focused on the undefined variable case.

Reply:
Sure, I'll update the comment to reflect the rule as it stands, and open a new issue for whether or not to tighten it up.